packages/puppeteer-core/src/bidi/Frame.ts

```diff
@@ -201,48 +210,32 @@ export class BidiFrame extends Frame {
   ): Promise<BidiHTTPResponse | null> {
     const {
       waitUntil = 'load',
-      timeout = this.#timeoutSettings.navigationTimeout(),
+      timeout: ms = this.#timeoutSettings.navigationTimeout(),
     } = options;
 
     const waitUntilEvent = lifeCycleToSubscribedEvent.get(
       getWaitUntilSingle(waitUntil)
     ) as string;
 
-    const [info] = await Deferred.race([
-      // TODO(lightning00blade): Should also keep tack of
-      // navigationAborted and navigationFailed
-      Promise.all([
-        waitForEvent<Bidi.BrowsingContext.NavigationInfo>(
+    const info = await firstValueFrom(
+      merge(
+        fromEvent(
           this.#context,
-          waitUntilEvent,
-          () => {
-            return true;
-          },
-          timeout,
-          this.#abortDeferred.valueOrThrow()
+          Bidi.ChromiumBidi.BrowsingContext.EventNames.NavigationStarted
+        ).pipe(
+          switchMap(() => {
```

Comment:
how does it work? would it wait for NavigationStarted and the waitUntilEvent event after that? or would it be more like the promise.all in the previous version

Reply:
the spec guarantees that that the waitUntil will happen after the nav started, right?

Reply:
Looked at the HTML spec the navigation started event is emitted very early on the navigation lifecycle before even the `HTML's` response has completed.